In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Step - 2 : Extract Data Understanding
**extract the data from data source**

**load the raw data**

**data understanding**

**data exploration**

# Step - 2 : Data Understanding
# Load Data & Understand every variable

In [4]:
df = pd.read_excel(r"E:\Gen_AI\AI & ML Notes by SRK Sir-20260428T155042Z-3-001\AI & ML Notes by SRK Sir\04. Machine Learning\Regression\Code & Datasets\insurance.xlsx")
df.head()

,age,sex,bmi,children,smoker,region,expenses
0,19,female,27.9,0,yes,southwest,16884.92
1,18,male,33.8,1,no,southeast,1725.55
2,28,male,33.0,3,no,southeast,4449.46
3,33,male,22.7,0,no,northwest,21984.47
4,32,male,28.9,0,no,northwest,3866.86


In [5]:
df.shape

(1338, 7)

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   str    
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   str    
 5   region    1338 non-null   str    
 6   expenses  1338 non-null   float64
dtypes: float64(2), int64(2), str(3)
memory usage: 73.3 KB


# Exploratory Data Analysis

In [8]:
df.dtypes

age           int64
sex             str
bmi         float64
children      int64
smoker          str
region          str
expenses    float64
dtype: object

In [9]:
d_types = dict(df.dtypes)
continous=[]
discreat=[]
count=[]
for i in d_types.items():
    if i[1]=='float64':
        continous.append(i[0])
    elif i[1]=='str':
        discreat.append(i[0])
    else:
        count.append(i[0])
print('continous:',continous)
print('discreat:',discreat)
print('count:',count)

continous: ['bmi', 'expenses']
discreat: ['sex', 'smoker', 'region']
count: ['age', 'children']


In [10]:

for i in df.columns.tolist():
    print(i, "\n", df[i].unique())

age 
 [19 18 28 33 32 31 46 37 60 25 62 23 56 27 52 30 34 59 63 55 22 26 35 24
 41 38 36 21 48 40 58 53 43 64 20 61 44 57 29 45 54 49 47 51 42 50 39]
sex 
 <StringArray>
['female', 'male']
Length: 2, dtype: str
bmi 
 [27.9 33.8 33.  22.7 28.9 25.7 33.4 27.7 29.8 25.8 26.2 26.3 34.4 39.8
 42.1 24.6 30.8 23.8 40.3 35.3 36.  32.4 34.1 31.9 28.  23.1 32.8 17.4
 36.3 35.6 28.6 28.3 36.4 20.4 20.8 36.7 39.9 26.6 36.6 21.8 37.1 37.3
 38.7 34.8 24.5 35.2 33.6 28.7 37.  31.8 31.7 22.9 27.4 33.7 24.7 25.9
 22.4 39.1 36.2 24.  24.8 28.5 28.1 32.  34.  29.6 35.5 26.9 38.3 37.6
 41.2 31.2 27.2 27.  39.5 31.3 20.  19.3 31.6 25.5 30.1 29.9 27.5 28.4
 30.9 35.1 29.7 35.7 32.2 49.1 23.4 29.  31.4 33.9 28.8 37.4 17.8 34.7
 26.5 22.  35.9 25.6 25.2 32.5 25.3 38.8 30.5 37.7 24.1 24.4 27.8 36.9
 39.6 28.2 33.2 18.9 41.5 30.3 16.  33.3 29.2 26.4 30.7 41.9 32.1 30.6
 18.1 39.3 36.1 22.3 26.7 23.2 38.1 22.2 38.4 29.1 22.1 26.8 30.  20.9
 17.3 34.2 25.4 40.2 24.3 42.4 19.8 30.2 29.4 27.6 20.6 21.6 40.6 48.1
 3

In [11]:
df[continous].describe()

,bmi,expenses
count,1338.000000,1338.000000
mean,30.665471,13270.422414
std,6.098382,12110.011240
min,16.000000,1121.870000
25%,26.300000,4740.287500
50%,30.400000,9382.030000
75%,34.700000,16639.915000
max,53.100000,63770.430000


In [12]:
df[discreat].describe()

,sex,smoker,region
count,1338,1338,1338
unique,2,2,4
top,male,no,southeast
freq,676,1064,364


In [13]:
df[count].describe()

,age,children
count,1338.000000,1338.000000
mean,39.207025,1.094918
std,14.049960,1.205493
min,18.000000,0.000000
25%,27.000000,0.000000
50%,39.000000,1.000000
75%,51.000000,2.000000
max,64.000000,5.000000


# step-3 data preprocessing

In [15]:
df.isnull().sum()

age         0
sex         0
bmi         0
children    0
smoker      0
region      0
expenses    0
dtype: int64

In [16]:
df.duplicated().sum()

1

In [17]:
df[df.duplicated(keep=False)]

,age,sex,bmi,children,smoker,region,expenses
195,19,male,30.6,0,no,northwest,1639.56
581,19,male,30.6,0,no,northwest,1639.56


In [18]:
df.drop_duplicates(inplace=True)

In [19]:
df.shape

(1337, 7)

# encoding 
# label or dummy if a column consist of two values we can use either of one

In [21]:
df.groupby("region")["expenses"].describe().T


region,northeast,northwest,southeast,southwest
count,324.000000,324.000000,364.000000,325.000000
mean,13406.384691,12450.840648,14735.411538,12346.937908
std,11255.802916,11073.125643,13971.098797,11557.179033
min,1694.800000,1621.340000,1121.870000,1241.570000
25%,5194.322500,4733.637500,4440.885000,4751.070000
50%,10057.655000,8976.980000,9294.130000,8798.590000
75%,16687.365000,14788.745000,19526.287500,13462.520000
max,58571.070000,60021.400000,63770.430000,52590.830000


In [22]:
df

,age,sex,bmi,children,smoker,region,expenses
0,19,female,27.9,0,yes,southwest,16884.92
1,18,male,33.8,1,no,southeast,1725.55
2,28,male,33.0,3,no,southeast,4449.46
3,33,male,22.7,0,no,northwest,21984.47
4,32,male,28.9,0,no,northwest,3866.86
...,...,...,...,...,...,...,...
1333,50,male,31.0,3,no,northwest,10600.55
1334,18,female,31.9,0,no,northeast,2205.98
1335,18,female,36.9,0,no,southeast,1629.83
1336,21,female,25.8,0,no,southwest,2007.95


In [23]:
df=pd.get_dummies(df,drop_first=True)

In [24]:
df

,age,bmi,children,expenses,sex_male,smoker_yes,region_northwest,region_southeast,region_southwest
0,19,27.9,0,16884.92,False,True,False,False,True
1,18,33.8,1,1725.55,True,False,False,True,False
2,28,33.0,3,4449.46,True,False,False,True,False
3,33,22.7,0,21984.47,True,False,True,False,False
4,32,28.9,0,3866.86,True,False,True,False,False
...,...,...,...,...,...,...,...,...,...
1333,50,31.0,3,10600.55,True,False,True,False,False
1334,18,31.9,0,2205.98,False,False,False,False,False
1335,18,36.9,0,1629.83,False,False,False,True,False
1336,21,25.8,0,2007.95,False,False,False,False,True


In [25]:
X = df.drop('expenses', axis=1)
y = df['expenses']


In [26]:
X

,age,bmi,children,sex_male,smoker_yes,region_northwest,region_southeast,region_southwest
0,19,27.9,0,False,True,False,False,True
1,18,33.8,1,True,False,False,True,False
2,28,33.0,3,True,False,False,True,False
3,33,22.7,0,True,False,True,False,False
4,32,28.9,0,True,False,True,False,False
...,...,...,...,...,...,...,...,...
1333,50,31.0,3,True,False,True,False,False
1334,18,31.9,0,False,False,False,False,False
1335,18,36.9,0,False,False,False,True,False
1336,21,25.8,0,False,False,False,False,True


In [27]:
y


0       16884.92
1        1725.55
2        4449.46
3       21984.47
4        3866.86
          ...   
1333    10600.55
1334     2205.98
1335     1629.83
1336     2007.95
1337    29141.36
Name: expenses, Length: 1337, dtype: float64

In [28]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=10)

# for regression project no need of scaling

# by applying scaling on i/p variables within regression project

#       -- no difference in R2 and RMSE


# by applying scaling on i/p variables and o/p variable within regression project

#       -- automatically RMSE decreases

#       -- R2 remains constant


#       Conclusion : no scaling for regression project

# linear regression

In [32]:
#model
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train,y_train)

print("coefficient:",model.coef_)
print("intercept:",model.intercept_)

# prediction on train data
#ypred_train=yhat calucation
ypred_train = model.predict(X_train)

#Evaluation on train data
from sklearn.metrics import mean_squared_error,r2_score
mse = mean_squared_error(y_train,ypred_train)
print("Train MSE:",mse)
print("Train RMSE:",np.sqrt(mse))
print("train R2:", r2_score(y_train, ypred_train))


# prediction on test data

ypred_test = model.predict(X_test)

#Evaluation on test data
from sklearn.metrics import mean_squared_error
mse = mean_squared_error(y_test,ypred_test)
print("Test RMSE:",np.sqrt(mse))
print("test R2:", r2_score(y_test, ypred_test))


from sklearn.model_selection import cross_val_score
cross_val_score(model, X_train,y_train,cv=5,scoring='r2').mean()

coefficient: [  255.35225841   338.4849792    515.69809747  -284.88168851
 24164.79047205  -240.5697816   -997.88563627  -682.12928289]
intercept: -11991.485523678208
Train MSE: 36706444.844729714
Train RMSE: 6058.584392804124
train R2: 0.7542086710880842
Test RMSE: 5997.974813586831
test R2: 0.7342269753811511


0.7446414960755954

In [33]:
#-ve value for categorical values

y = β₁x₁ + β₀
y=100(1)+50
y=100(2)+50
delatychange=100

# Non Linear Regression

In [36]:
from sklearn.preprocessing import PolynomialFeatures
polynomial_converter = PolynomialFeatures(degree=2,include_bias=False)
X_train_poly = pd.DataFrame(polynomial_converter.fit_transform(X_train))
X_train_poly

from sklearn.linear_model import LinearRegression
nlr_model = LinearRegression()
nlr_model.fit(X_train_poly,y_train)

print("Intercept:",nlr_model.intercept_)
print("coefficients:",nlr_model.coef_)

#prediction
ypred_train = nlr_model.predict(X_train_poly)

#Train R2
from sklearn.metrics import mean_absolute_error,mean_squared_error,r2_score
print("Train R2:",r2_score(y_train,ypred_train))
mse = mean_squared_error(y_train,ypred_train)
print("Train MSE:",mse)
print("Train RMSE:",np.sqrt(mse))

#Cross Validation Score
from sklearn.model_selection import cross_val_score
print("CV Score:",cross_val_score(nlr_model,X_train_poly,y_train,cv=5).mean())

#Test R2
X_test_poly = pd.DataFrame(polynomial_converter.transform(X_test))
ypred_test = nlr_model.predict(X_test_poly)
print("Test R2:",r2_score(y_test,ypred_test))


Intercept: -1506.7010712625342
coefficients: [-1.47839118e+02  4.21926081e+02  1.25134215e+03 -6.73486764e+02
 -1.05710750e+04  7.72335144e+02  8.47497833e+02 -3.64287023e+02
  4.59459972e+00  8.15240146e-01 -8.71943486e-01  4.94382570e+00
 -2.81384626e+00 -8.98622693e+00  3.92909125e+01  3.74318322e+01
 -6.47270636e+00  2.88276022e+00  2.70217038e+01  1.52297285e+03
 -7.62049054e+01 -1.34347260e+02 -5.29134047e+01 -5.98857817e+01
 -1.57400093e+02 -6.35833640e+02  6.60524206e+01 -2.06471882e+02
 -5.17671820e+02 -6.73486764e+02 -1.22507470e+03 -9.48879881e+00
  4.02329676e+02  7.19447623e+01 -1.05710750e+04  1.01768754e+03
 -1.62780995e+03  1.52181532e+03  7.72335144e+02  0.00000000e+00
  0.00000000e+00  8.47497833e+02  0.00000000e+00 -3.64287023e+02]
Train R2: 0.8580421400288407
Train MSE: 21199968.202192374
Train RMSE: 4604.342320265987
CV Score: 0.8447977365105375
Test R2: 0.7910233367181556


In [37]:
#grid search cv
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import Lasso

estimator = Lasso()

param_grid = {"alpha": list(range(1,101))}    # ← empty, needs values

model_hp = GridSearchCV(estimator, param_grid, cv=5, scoring='r2')

model_hp.fit(X_train, y_train)

model_hp.best_params_

{'alpha': 77}

In [38]:
# Modelling
from sklearn.linear_model import Lasso
lasso_best = Lasso(alpha=91)
lasso_best.fit(X_train, y_train)

print("Intercept:",lasso_best.intercept_)
print("coefficients:",lasso_best.coef_)

# prediction on train data
ypred_train = lasso_best.predict(X_train)

# Evaluation on train data
from sklearn.metrics import mean_squared_error, r2_score
print("Train R2:", r2_score(y_train, ypred_train))

# prediction on test data
ypred_test = lasso_best.predict(X_test)

# Evaluation
print("test R2:", r2_score(y_test, ypred_test))

from sklearn.model_selection import cross_val_score
print("Cross Validation Score:", cross_val_score(lasso_best, X, y, cv=5).mean())

Intercept: -11884.788761349211
coefficients: [  255.59284877   321.24452199   456.66336073    -0.
 23579.17285623     0.          -152.60049498    -0.        ]
Train R2: 0.7529786949108372
test R2: 0.7349206017547216
Cross Validation Score: 0.7462817899945062


# Hyper parameter seacrch for ridge search 

In [40]:
#if the range value if it has max value constant then lasso is not useful 1,100  next 100..200 then range stop at 199
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import Ridge

estimator = Ridge()

param_grid = {"alpha": list(range(1,100))}    # ← empty, needs values

model_hp = GridSearchCV(estimator, param_grid, cv=5, scoring='r2')

model_hp.fit(X_train, y_train)

model_hp.best_params_

{'alpha': 2}

In [41]:
# Modelling
from sklearn.linear_model import Ridge
ridge_best = Ridge(alpha=2)
ridge_best.fit(X_train, y_train)

print("Intercept:",ridge_best.intercept_)
print("coefficients:",ridge_best.coef_)

# prediction on train data
ypred_train = ridge_best.predict(X_train)

# Evaluation on train data
from sklearn.metrics import mean_squared_error, r2_score
print("Train R2:", r2_score(y_train, ypred_train))

# prediction on test data
ypred_test = ridge_best.predict(X_test)

# Evaluation
print("test R2:", r2_score(y_test, ypred_test))

from sklearn.model_selection import cross_val_score
print("Cross Validation Score:", cross_val_score(ridge_best, X, y, cv=5).mean())

Intercept: -11928.337437599452
coefficients: [  255.08551128   337.93180121   517.19507669  -263.95597408
 23889.45867045  -237.9703293   -971.9231864   -675.66494552]
Train R2: 0.7541258275300812
test R2: 0.7351587454810091
Cross Validation Score: 0.746724565741593


# ElasticNet

In [43]:
from sklearn.model_selection import GridSearchCV

# model
from sklearn.linear_model import ElasticNet
estimator = ElasticNet()

# parameters & values
param_grid = {"alpha":list(range(1,101)),
              "l1_ratio":[0,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1]}

#Identifying the best value of the parameter within given values for the given
model_hp = GridSearchCV(estimator,param_grid,cv=5,scoring='r2')

model_hp.fit(X_train,y_train)

model_hp.best_params_

C:\Users\abhis\anaconda3\envs\tf215\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:716: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 5.149e+10, tolerance: 1.293e+07
Linear regression models with a zero l1 penalization strength are more efficiently fitted using one of the solvers implemented in sklearn.linear_model.Ridge/RidgeCV instead.
  model = cd_fast.enet_coordinate_descent(
C:\Users\abhis\anaconda3\envs\tf215\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:716: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 4.754e+10, tolerance: 1.188e+07
Linear regression models with a zero l1 penalization strength are more efficiently fitted using one of the solvers implemented in sklearn.linea

{'alpha': 77, 'l1_ratio': 1}

In [44]:
#Modelling
enr_best = ElasticNet(alpha=77,l1_ratio=1)
enr_best.fit(X_train,y_train)

print("Intercept:",enr_best.intercept_)
print("coefficients:",enr_best.coef_)

#predictions
train_predictions = enr_best.predict(X_train)
test_predictions = enr_best.predict(X_test)

#Evaluation
print("Train R2:",enr_best.score(X_train,y_train))
print("Test R2:",enr_best.score(X_test,y_test))

from sklearn.model_selection import cross_val_score
print("Cross Validation Score:",cross_val_score(enr_best,X,y,cv=5).mean())

Intercept: -11950.385901787833
coefficients: [  255.58858994   323.20646857   465.31951264    -0.
 23668.30006322     0.          -234.05194806    -0.        ]
Train R2: 0.7531825331482958
Test R2: 0.7346650524233063
Cross Validation Score: 0.7463616040263634


In [45]:
#train r2 should be equal to test r2
#train r2 should be equal to cv
#which is having highest test r2 value

══════════════════════════════════════════════════════════════════════
  MODEL COMPARISON — Regression Results
══════════════════════════════════════════════════════════════════════

┌─────────────┬──────────┬──────────┬──────────┬────────────────┐
│ Model       │ Train R² │ Test R²  │ CV Score │ Train-Test Gap │
├─────────────┼──────────┼──────────┼──────────┼────────────────┤
│ Non-linear  │  0.8580  │  0.7910  │  0.8448  │     0.0670     │ ★ BEST
│ Ridge       │  0.7541  │  0.7352  │  0.7467  │     0.0190     │
│ Lasso       │  0.7530  │  0.7349  │  0.7463  │     0.0181     │
│ Elastic Net │  0.7532  │  0.7347  │  0.7464  │     0.0185     │
│ Linear      │  0.7542  │  0.7342  │  0.7446  │     0.0200     │
└─────────────┴──────────┴──────────┴──────────┴────────────────┘

──────────────────────────────────────────────────────────────────────
  OBSERVATIONS
──────────────────────────────────────────────────────────────────────

  → Non-linear wins on Test R² by ~5.6 percentage points
  → CV Score (0.8448) confirms it generalizes, not lucky split
  → Linear / Lasso / Ridge / Elastic are tied (within 0.001 on Test)
  → Regularization didn't help → base linear wasn't overfitting
  → Data has non-linear patterns that linear models can't capture

──────────────────────────────────────────────────────────────────────
  DECISION
──────────────────────────────────────────────────────────────────────

  ✓ Use Non-linear model
  ⚠ Verify stability with multiple random_state values
  ⚠ Train-Test gap (0.067) is larger but acceptable given CV stability

══════════════════════════════════════════════════════════════════════